fdb-relational-core/src/main/java/com/apple/foundationdb/relational/recordlayer/query/visitors/QueryVisitor.java

```diff
@@ -79,18 +80,46 @@ public static QueryVisitor of(@Nonnull BaseVisitor baseVisitor) {
         return new QueryVisitor(baseVisitor);
     }
 
+    /**
+     * Capture semantic type structure from expressions.
+     *
+     * This preserves struct type names (like "STRUCT_1", "STRUCT_2") that are known during
+     * semantic analysis but get lost in planner Type conversion (which has null names).
+     *
+     * Field names in the returned StructType are TEMPORARY PLACEHOLDERS. The actual field names
+     * will be taken from the planner Type.Record during result set creation.
+     *
+     * @param expressions The expressions from LogicalOperator.getOutput()
+     * @return List of DataTypes preserving struct type names (field names are placeholders)
+     */
+    @Nonnull
+    private static List<DataType> captureSemanticTypeStructure(
```

Comment:
`Expressions` is a comprehension that captures operations on an ordered list of `Expression`s, please move the method and make it part of it, this enables for example, maintaining and caching the underlying data types. Something like:

```java
Iterable<DataType> getDataTypes() {
    return ....
}
```